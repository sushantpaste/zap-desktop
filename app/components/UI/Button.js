import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Button as BaseButton, Flex, Text } from 'rebass'
import Spinner from './Spinner'

const Wrapper = styled(BaseButton)`
  transition: all 0.25s;
  outline: none;
  border-radius: 5;
  font-weight: normal;
  line-height: '18px';
  &:disabled {
    opacity: 0.5;
  }
  &:hover:enabled {
    cursor: pointer;
  }
`
Wrapper.displayName = 'Button'

/**
 * @render react
 * @name Button
 * @example
 * <Button><Basic button</Button>
 */
class Button extends React.PureComponent {
  static displayName = 'Button'
  static defaultProps = {
    processing: false,
    size: 'medium',
    variant: 'normal'
  }
  static propTypes = {
    processing: PropTypes.bool,
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    variant: PropTypes.string
  }

  render() {
    const { children, processing, size, ...rest } = this.props
    const sizes = {
      small: {
        x: 3,
        y: 2,
        fontSize: 's'
      },
      medium: {
        x: 3,
        y: 2,
        fontSize: 'm'
      },
      large: {
        x: 5,
        y: 3,
        fontSize: 'l'
      }
    }
    return (
      <Wrapper
        px={sizes[size]['x']}
        py={sizes[size]['y']}
        fontSize={sizes[size]['fontSize']}
        {...rest}
      >
        {processing ? (
          <Flex>
            {processing && <Spinner size="2em" mr="0.5em" />}
            <Text fontFamily="sans">{children}</Text>
          </Flex>
        ) : (
          <Text fontFamily="sans">{children}</Text>
        )}
      </Wrapper>
    )
  }
}

export default Button
