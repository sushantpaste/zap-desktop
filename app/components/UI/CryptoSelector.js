import React from 'react'
import PropTypes from 'prop-types'
import { Box, Flex } from 'rebass'
import styled, { withTheme } from 'styled-components'
import FaAngleDown from 'react-icons/lib/fa/angle-down'
import FaAngleUp from 'react-icons/lib/fa/angle-up'
import Check from 'components/Icon/Check'
import Text from 'components/UI/Text'

/**
 * Container
 */
const CryptoSelectorContainer = styled(Flex)({})
CryptoSelectorContainer.defaultProps = {
  flexDirection: 'column',
  flexWrap: 'none',
  display: 'relative'
}

/**
 * Button
 */
const CryptoSelectorButton = styled(Box)({
  appearance: 'none',
  display: 'inline-block',
  textAlign: 'center',
  lineHeight: 'inherit',
  textDecoration: 'none',
  border: 'none',
  outline: 'none',
  background: 'transparent',
  color: 'inherit',
  cursor: 'pointer'
})
CryptoSelectorButton.defaultProps = {
  as: 'button',
  m: 0,
  px: 0,
  py: 2,
  textAlign: 'left'
}

/**
 * Menu
 */
const MenuContainer = styled(Box)({
  display: 'relative'
})

const Menu = styled(Box)({
  cursor: 'pointer',
  display: 'inline-block',
  position: 'absolute',
  'z-index': '999',
  'list-style-type': 'none',
  'border-radius': '3px',
  'box-shadow': '0 3px 4px 0 rgba(30, 30, 30, 0.5)'
})
Menu.defaultProps = {
  as: 'ul',
  m: 0,
  p: 0,
  bg: 'lightestBackground'
}

/**
 * MenuItem
 */
const MenuItem = styled(Box)`
  cursor: pointer;
  &:hover {
    background-color: ${props => props.theme.colors.darkestBackground};
  }
`
MenuItem.defaultProps = {
  as: 'li',
  px: 2,
  py: 2
}

/**
 * @render react
 * @name CryptoSelector
 * @example
 * <CryptoSelector />
 */
class CryptoSelector extends React.Component {
  state = {
    isOpen: false
  }
  onChange = this.onChange.bind(this)
  toggleMenu = this.toggleMenu.bind(this)
  setWrapperRef = this.setWrapperRef.bind(this)
  handleClickOutside = this.handleClickOutside.bind(this)

  static propTypes = {
    currency: PropTypes.string.isRequired,
    currencies: PropTypes.arrayOf(
      PropTypes.shape({
        key: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired
      })
    ).isRequired,
    onChange: PropTypes.func,
    setCurrency: PropTypes.func.isRequired
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside)
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside)
  }

  onChange(key) {
    const { onChange, setCurrency, currency } = this.props
    if (key !== currency) {
      setCurrency(key)
      if (onChange) {
        onChange(key)
      }
    }
    this.setState({ isOpen: false })
  }

  setWrapperRef(node) {
    this.wrapperRef = node
  }

  handleClickOutside(event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({ isOpen: false })
    }
  }

  toggleMenu() {
    const { isOpen } = this.state
    this.setState({ isOpen: !isOpen })
  }

  render() {
    const { isOpen } = this.state
    const { currency, currencies, theme, ...rest } = this.props
    const selectedCurrency = currencies.find(c => c.key === currency)
    return (
      <CryptoSelectorContainer ref={this.setWrapperRef} {...rest}>
        <CryptoSelectorButton type="button" onClick={this.toggleMenu}>
          <Text textAlign="left">
            {selectedCurrency ? selectedCurrency.name : currency}{' '}
            {isOpen ? <FaAngleUp /> : <FaAngleDown />}
          </Text>
        </CryptoSelectorButton>
        {isOpen && (
          <MenuContainer>
            <Menu>
              {currencies.map(item => {
                return (
                  <MenuItem key={item.key} onClick={() => this.onChange(item.key)}>
                    <Flex alignItems="center">
                      <Text width="18px">
                        {currency === item.key && (
                          <Check height="0.95em" color={theme.colors.superGreen} />
                        )}
                      </Text>
                      <Text>{item.name}</Text>
                    </Flex>
                  </MenuItem>
                )
              })}
            </Menu>
          </MenuContainer>
        )}
      </CryptoSelectorContainer>
    )
  }
}

export default withTheme(CryptoSelector)
