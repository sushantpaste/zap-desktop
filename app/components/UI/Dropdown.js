import React from 'react'
import PropTypes from 'prop-types'
import { Box, Flex } from 'rebass'
import styled, { withTheme } from 'styled-components'
import FaAngleDown from 'react-icons/lib/fa/angle-down'
import FaAngleUp from 'react-icons/lib/fa/angle-up'
import Check from 'components/Icon/Check'
import Text from 'components/UI/Text'

/**
 * Container
 */
const DropdownContainer = styled(Flex)({})
DropdownContainer.defaultProps = {
  flexDirection: 'column',
  flexWrap: 'none',
  display: 'relative'
}

/**
 * Button
 */
const DropdownButton = styled(Box)({
  appearance: 'none',
  display: 'inline-block',
  textAlign: 'center',
  lineHeight: 'inherit',
  textDecoration: 'none',
  border: 'none',
  outline: 'none',
  background: 'transparent',
  color: 'inherit',
  cursor: 'pointer'
})
DropdownButton.defaultProps = {
  as: 'button',
  m: 0,
  px: 0,
  py: 2,
  textAlign: 'left'
}

/**
 * Menu
 */
const MenuContainer = styled(Box)({
  display: 'relative'
})

const Menu = styled(Box)({
  cursor: 'pointer',
  display: 'inline-block',
  position: 'absolute',
  'z-index': '999',
  'min-width': '70px',
  'list-style-type': 'none',
  'border-radius': '3px',
  'box-shadow': '0 3px 4px 0 rgba(30, 30, 30, 0.5)'
})
Menu.defaultProps = {
  as: 'ul',
  m: 0,
  p: 0,
  bg: 'lightestBackground'
}

/**
 * MenuItem
 */
const MenuItem = styled(Box)`
  cursor: pointer;
  &:hover {
    background-color: ${props => props.theme.colors.darkestBackground};
  }
`
MenuItem.defaultProps = {
  as: 'li',
  px: 2,
  py: 2
}

/**
 * @render react
 * @name Dropdown
 * @example
 * <Dropdown items={[
 *    {name: 'Item 1', key: 'key1'},
 *    {name: 'Item 2', key: 'key2'}
 *  ]} activeKey="key1" />
 */
class Dropdown extends React.Component {
  state = {
    isOpen: false
  }
  onChange = this.onChange.bind(this)
  toggleMenu = this.toggleMenu.bind(this)
  setWrapperRef = this.setWrapperRef.bind(this)
  handleClickOutside = this.handleClickOutside.bind(this)

  static propTypes = {
    activeKey: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(
      PropTypes.shape({
        key: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired
      })
    ).isRequired,
    onChange: PropTypes.func
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside)
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside)
  }

  onChange(key) {
    const { onChange, activeKey } = this.props
    if (key !== activeKey) {
      if (onChange) {
        onChange(key)
      }
    }
    this.setState({ isOpen: false })
  }

  setWrapperRef(node) {
    this.wrapperRef = node
  }

  handleClickOutside(event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({ isOpen: false })
    }
  }

  toggleMenu() {
    const { isOpen } = this.state
    this.setState({ isOpen: !isOpen })
  }

  render() {
    const { isOpen } = this.state
    const { activeKey, items, theme, ...rest } = this.props
    const selectedItem = items.find(c => c.key === activeKey)
    return (
      <DropdownContainer ref={this.setWrapperRef} {...rest}>
        <DropdownButton type="button" onClick={this.toggleMenu}>
          <Text textAlign="left">
            {selectedItem ? selectedItem.name : activeKey}{' '}
            {isOpen ? <FaAngleUp /> : <FaAngleDown />}
          </Text>
        </DropdownButton>
        {isOpen && (
          <MenuContainer>
            <Menu>
              {items.map(item => {
                return (
                  <MenuItem key={item.key} onClick={() => this.onChange(item.key)}>
                    <Flex alignItems="center">
                      <Text width="18px">
                        {activeKey === item.key && (
                          <Check height="0.95em" color={theme.colors.superGreen} />
                        )}
                      </Text>
                      <Text>{item.name}</Text>
                    </Flex>
                  </MenuItem>
                )
              })}
            </Menu>
          </MenuContainer>
        )}
      </DropdownContainer>
    )
  }
}

export default withTheme(Dropdown)
