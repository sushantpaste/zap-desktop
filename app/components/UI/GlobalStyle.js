import { createGlobalStyle } from 'styled-components'

const GlobalStyle = createGlobalStyle`
  html {
    box-sizing: border-box;
  }
  *, *:before, *:after {
    box-sizing: inherit;
  }
  body {
    position: relative;
    overflow-y: hidden;
    -webkit-font-smoothing: antialiased;
    -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
    font-family: 'Roboto', Arial, Helvetica, sans-serif;
    font-size: 13px;
  }
`

export default GlobalStyle
