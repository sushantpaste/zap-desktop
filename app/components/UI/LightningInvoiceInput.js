import React from 'react'
import PropTypes from 'prop-types'
import { asField } from 'informed'
import { isOnchain, isLn } from 'lib/utils/crypto'
import TextArea from 'components/UI/TextArea'
import FormFieldMessage from 'components/UI/FormFieldMessage'

/**
 * @render react
 * @name LightningInvoiceInput
 * @example
 * <LightningInvoiceInput
     network="testnet"
     field="testnet"
     id="testnet"
     validateOnChange
     validateOnBlur />
 */
class LightningInvoiceInput extends React.Component {
  static displayName = 'LightningInvoiceInput'

  static propTypes = {
    required: PropTypes.bool,
    chain: PropTypes.oneOf(['bitcoin', 'litecoin']),
    network: PropTypes.oneOf(['mainnet', 'testnet', 'regtest'])
  }

  static defaultProps = {
    required: false
  }

  validate = value => {
    const { network, chain, required } = this.props
    if (required && (!value || value.trim() === '')) {
      return 'This is a required field'
    }
    if (value && !isLn(value, chain, network) && !isOnchain(value, chain, network)) {
      return 'Not a valid address.'
    }
  }

  render() {
    return (
      <InformedTextArea
        placeholder="Paste a Lightning Payment Request or Bitcoin Address here"
        rows={5}
        {...this.props}
        validate={this.validate}
      />
    )
  }
}

const InformedTextArea = asField(({ fieldState, fieldApi, ...props }) => {
  const { value } = fieldState
  const { chain, network, ...rest } = props
  return (
    <React.Fragment>
      <TextArea {...rest} />
      {value &&
        !fieldState.error && (
          <FormFieldMessage variant="success" mt={2}>
            Valid {isLn(value, chain, network) ? 'lightning' : chain} address{' '}
            {network !== 'mainnet' && `(${network})`}
          </FormFieldMessage>
        )}
    </React.Fragment>
  )
})

export default LightningInvoiceInput
