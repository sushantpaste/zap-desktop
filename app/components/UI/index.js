export BackgroundDark from './BackgroundDark'
export BackgroundLight from './BackgroundLight'
export BackgroundLightest from './BackgroundLightest'
export Bar from './Bar'
export Button from './Button'
export FormFieldMessage from './FormFieldMessage'
export GlobalStyle from './GlobalStyle'
export Heading from './Heading'
export Input from './Input'
export Label from './Label'
export LightningInvoiceInput from './LightningInvoiceInput'
export MainContent from './MainContent'
export Menu from './Menu'
export MenuItem from './MenuItem'
export MenuItemGroup from './MenuItemGroup'
export Modal from './Modal'
export Notification from './Notification'
export Page from './Page'
export Range from './Range'
export Select from './Select'
export Sidebar from './Sidebar'
export Spinner from './Spinner'
export Text from './Text'
export TextArea from './TextArea'
export Titlebar from './Titlebar'
export Toggle from './Toggle'
