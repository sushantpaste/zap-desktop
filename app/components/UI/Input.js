import React from 'react'
import { asField } from 'informed'
import { withTheme } from 'styled-components'
import system from '@rebass/components'
import { styles } from 'styled-system'
import { Flex } from 'rebass'
import { FormFieldMessage } from 'components/UI'

// Create an html input element that accepts all style props from styled-system.
const SystemInput = system(
  {
    as: 'input',
    border: 1,
    borderColor: 'gray',
    borderRadius: 5,
    bg: 'transparent',
    color: 'primaryText',
    fontFamily: 'sans',
    fontSize: 'm',
    fontWeight: 'light',
    p: 3,
    width: 1
  },
  ...Object.keys(styles)
)

/**
 * @render react
 * @name Input
 * @example
 * <Input />
 */
class Input extends React.Component {
  static displayName = 'Input'

  state = {
    hasFocus: false
  }

  constructor(props) {
    super(props)
    const { forwardedRef } = this.props
    this.inputRef = forwardedRef || React.createRef()
  }

  render() {
    const {
      css,
      onChange,
      onBlur,
      onFocus,
      initialValue,
      forwardedRef,
      theme,
      fieldApi,
      fieldState,
      justifyContent,
      ...rest
    } = this.props
    const { readOnly } = this.props
    const { hasFocus } = this.state
    const { setValue, setTouched } = fieldApi
    const { value } = fieldState
    const isValid = value && !fieldState.error

    // Calculate the border color based on the current field state.
    let borderColor
    if (readOnly) {
      borderColor = theme.colors.gray
    } else if (fieldState.error) {
      borderColor = theme.colors.superRed
    } else if (value && !fieldState.error) {
      borderColor = theme.colors.superGreen
    }

    return (
      <Flex flexDirection="column" justifyContent={justifyContent}>
        <SystemInput
          borderColor={borderColor || theme.colors.gray}
          css={Object.assign(
            {
              outline: 'none',
              '&:not([readOnly]):not([disabled]):focus': {
                border: `1px solid ${
                  isValid ? theme.colors.superGreen : theme.colors.lightningOrange
                } }`
              }
            },
            css
          )}
          {...rest}
          ref={this.inputRef}
          value={!value && value !== 0 ? '' : initialValue || value}
          onChange={e => {
            setValue(e.target.value)
            if (onChange) {
              onChange(e)
            }
          }}
          onBlur={e => {
            setTouched()
            // Make the state aware that the element is now focused.
            const newHasFocus = document.activeElement === this.inputRef.current
            if (hasFocus !== newHasFocus) {
              this.setState({ hasFocus: newHasFocus })
            }
            if (onBlur) {
              onBlur(e)
            }
          }}
          onFocus={e => {
            // Make the state aware that the element is no longer focused.
            const newHasFocus = document.activeElement === this.inputRef.current
            if (hasFocus !== newHasFocus) {
              this.setState({ hasFocus: newHasFocus })
            }
            if (onFocus) {
              onFocus(e)
            }
          }}
          error={fieldState.error}
        />
        {fieldState.error && (
          <FormFieldMessage
            variant={hasFocus ? 'warning' : 'error'}
            justifyContent={justifyContent}
            mt={2}
          >
            {fieldState.error}
          </FormFieldMessage>
        )}
      </Flex>
    )
  }
}

export default asField(withTheme(Input))
