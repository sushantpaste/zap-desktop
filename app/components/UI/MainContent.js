import React from 'react'
import { BackgroundDark } from 'components/UI'

/**
 * @render react
 * @name MainContent
 * @example
 * <MainContent>Some content</MainContent>
 */
const MainContent = props => (
  <BackgroundDark as="article" width={1} p={3} css={{ height: '100%' }} {...props} />
)

export default MainContent
