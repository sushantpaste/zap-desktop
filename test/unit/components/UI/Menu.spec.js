import React from 'react'
import renderer from 'react-test-renderer'
import { dark } from 'themes'
import { ThemeProvider } from 'styled-components'
import { Menu, MenuItem, MenuItemGroup } from 'components/UI'

describe('component.UI.Menu', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={dark}>
          <Menu>
            <MenuItemGroup title="My Items">
              <MenuItem key="item1">My item 1</MenuItem>
              <MenuItem key="item2">My item 2</MenuItem>
              <MenuItem key="item3">My item 3</MenuItem>
            </MenuItemGroup>

            <MenuItemGroup title="Other items">
              <MenuItem key="item4">Other item 1</MenuItem>
              <MenuItem key="item5">Other item 2</MenuItem>
              <MenuItem key="item6">Other item 3</MenuItem>
            </MenuItemGroup>
          </Menu>
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
