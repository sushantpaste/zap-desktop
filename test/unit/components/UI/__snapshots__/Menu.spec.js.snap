// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component.UI.Menu should render correctly 1`] = `
.c0 {
  margin-bottom: 8px;
  padding: 0px;
  outline: none;
}

.c2 {
  padding-top: 8px;
  padding-bottom: 8px;
  list-style-type: none;
}

.c2.rc-menu-item-active {
  color: #fd9800;
  cursor: pointer;
}

.c2.rc-menu-item-selected {
  color: #fd9800;
}

.c1 {
  margin-bottom: 16px;
  list-style-type: none;
}

.c1 > .rc-menu-item-group-title {
  font-weight: bold;
  padding: 8px 0;
}

.c1 > .rc-menu-item-group-list {
  padding: 0;
}

<ul
  className="rc-menu c0 rc-menu-root rc-menu-vertical"
  mb={2}
  onKeyDown={[Function]}
  p={0}
  role="menu"
  style={Object {}}
  tabIndex="0"
>
  <li
    className="c1 rc-menu-item-group"
    mb={3}
  >
    <div
      className="rc-menu-item-group-title"
      title="My Items"
    >
      My Items
    </div>
    <ul
      className="rc-menu-item-group-list"
    >
      <li
        className="rc-menu-item c2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        py={2}
        role="menuitem"
        style={Object {}}
      >
        My item 1
      </li>
      <li
        className="rc-menu-item c2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        py={2}
        role="menuitem"
        style={Object {}}
      >
        My item 2
      </li>
      <li
        className="rc-menu-item c2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        py={2}
        role="menuitem"
        style={Object {}}
      >
        My item 3
      </li>
    </ul>
  </li>
  <li
    className="c1 rc-menu-item-group"
    mb={3}
  >
    <div
      className="rc-menu-item-group-title"
      title="Other items"
    >
      Other items
    </div>
    <ul
      className="rc-menu-item-group-list"
    >
      <li
        className="rc-menu-item c2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        py={2}
        role="menuitem"
        style={Object {}}
      >
        Other item 1
      </li>
      <li
        className="rc-menu-item c2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        py={2}
        role="menuitem"
        style={Object {}}
      >
        Other item 2
      </li>
      <li
        className="rc-menu-item c2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        py={2}
        role="menuitem"
        style={Object {}}
      >
        Other item 3
      </li>
    </ul>
  </li>
</ul>
`;
